File: Source/Aladdin.CAPI.ANSI.Java/src/aladdin/capi/ansi/keyx/rsa/pkcs1/Encoding.java
package aladdin.capi.ansi.keyx.rsa.pkcs1;
import aladdin.capi.*;
import java.io.*;

///////////////////////////////////////////////////////////////////////////
// Кодирование EME PKCS1
///////////////////////////////////////////////////////////////////////////
public class Encoding
{
    // закодировать данные
    public static byte[] encode(IRand rand, byte[] data, int k) throws IOException
    {
        // проверить размер данных
        if (data.length > k - 11) throw new IllegalArgumentException();

        // выделить память для результата
        byte[] encoded = new byte[k]; encoded[0] = 0x00; encoded[1] = 0x02;

        // сгенерировать случайные данные
        rand.generate(encoded, 2, k - data.length - 3); 

        // для всех сгенерированных данных
        for (int i = 2; i < k - data.length - 1; i++)
        {
            // удостовериться что байт ненулевой
            while (encoded[i] == 0) rand.generate(encoded, i, 1);
        }
        // скопировать данные
        System.arraycopy(data, 0, encoded, k - data.length, data.length); 

        // записать значение разделителя
        encoded[k - data.length - 1] = 0x00; return encoded; 
    }
    // раскодировать данные
    public static byte[] decode(byte[] encoded) throws IOException
    {
        // проверить первые байты
        if (encoded[0] != 0 || encoded[1] != 2) throw new IOException();

        // найти нулевой байт
        int i; for (i = 2; i < encoded.length && encoded[i] != 0; i++) {} 

        // проверить наличие нулевого байта
        if (i == encoded.length) throw new IOException();

        // выделить память для результата
        byte[] data = new byte[encoded.length - i - 1]; 

        // извлечь данные
        System.arraycopy(encoded, i + 1, data, 0, data.length); return data; 
    }
}

File: Source/Aladdin.CAPI.ANSI/CAPI/ANSI/Keyx/RSA/PKCS1/Encoding.cs
using System;
using System.IO;

namespace Aladdin.CAPI.ANSI.Keyx.RSA.PKCS1
{
    ///////////////////////////////////////////////////////////////////////////
    // Кодирование EME PKCS1
    ///////////////////////////////////////////////////////////////////////////
    public static class Encoding
    {
        // закодировать данные
		public static byte[] Encode(IRand rand, byte[] data, int k)
		{
		    // проверить размер данных
		    if (data.Length > k - 11) throw new InvalidDataException();

		    // выделить память для результата
		    byte[] encoded = new byte[k]; encoded[0] = 0x00; encoded[1] = 0x02;

		    // сгенерировать случайные данные
		    rand.Generate(encoded, 2, k - data.Length - 3); 

		    // для всех сгенерированных данных
		    for (int i = 2; i < k - data.Length - 1; i++)
		    {
			    // удостовериться что байт ненулевой
			    while (encoded[i] == 0) rand.Generate(encoded, i, 1);
		    }
		    // скопировать данные
		    Array.Copy(data, 0, encoded, k - data.Length, data.Length); 
		
		    // записать значение разделителя
		    encoded[k - data.Length - 1] = 0x00; return encoded; 
		}
        // раскодировать данные
		public static byte[] Decode(byte[] encoded)
		{
		    // проверить первые байты
		    if (encoded[0] != 0 || encoded[1] != 2) throw new InvalidDataException();

		    // найти нулевой байт
		    int i; for (i = 2; i < encoded.Length && encoded[i] != 0; i++); 

		    // проверить наличие нулевого байта
		    if (i == encoded.Length) throw new InvalidDataException();

		    // выделить память для результата
		    byte[] data = new byte[encoded.Length - i - 1]; 

		    // извлечь данные
		    Array.Copy(encoded, i + 1, data, 0, data.Length); return data; 
		}
    }
}
